### DISCUSSION TITLE
Latest Debian install follies: no llvm-NN-linker-tools

### DISCUSSION THREAD

**Discussion by RossBoylan** [dosemu2/dosemu2#1692] [General]
The bug tracker was seeming a little heavy-weight; let me know if this is not the best way to be reporting this stuff.

I tried the latest instructions for [building on Debian](https://github.com/dosemu2/dosemu2/wiki/Building-packages-for-Debian) and the latest source code.
1. Installed the one new tool, `checkinstall`.  That installed, but the previous install of `dosemu2` got removed since it was considered broken.
2. New comcom instructions went well.  @stsp great find on `checkinstall`!
3. New build of `dosemu2` failed early, at the build dependencies.  Neither `llvm-12-linker-tools` nor `llvm-13-linker-tools` seems to be in Debian stable.

LLVM 11 seems to be the latest in Debian stable (there's also 9).  There is a package `lld` "LLD is a new, high-performance linker. It is built as a set of reusable components which highly leverage existing libraries in the larger LLVM Project.".  But I don't see anything looking like linker-tools.  Things in bin for package `lld-11` have names like ld.lld, lld, lld-link, wasm-ld (often with variations). 

There are `liblld-NN-dev` packages, but the description is "This package provides the header files to build extension over lld", which is of no direct help.

If there are particular commands that are needed I can search for them.  LLVM up through 14 is in testing, and 15 is in sid/unstable.  I notice the version for 14 does have an `llvm-14-linker-tools` package.  Maybe that was introduced after version 11?

**Comment by RossBoylan** [comment]
I deleted the references to the `linker-tools` from `debian/control` and otherwise followed the instructions.  It seems to have worked.

There were some complaints from lintian, as before:
```
Now running lintian dosemu2_2.0~pre9-1_amd64.changes ...
E: dosemu2 changes: bad-distribution-in-changes-file focal
E: dosemu2: custom-library-search-path usr/lib/dosemu/libplugin_fdpp.so /usr/lib/fdpp
E: dosemu2: depends-on-essential-package-without-using-version Depends: bash
E: dosemu2: depends-on-essential-package-without-using-version Depends: coreutils
W: dosemu2: package-contains-upstream-installation-documentation usr/share/doc/dosemu2/INSTALL
Finished running lintian.
```
but otherwise no errors or warnings that I saw.  Is the custom library search path a problem?

At the final install step it did install `fluid-soundfont-gm` this time; perhaps the dependency failures in my earlier attempt stopped it then.  It's a bit large (120MB).

```shell
root@barley:/usr/local/src/dos# date; apt install ./dosemu2_2.0~pre9-1_amd64.deb
Sat 18 Jun 2022 04:35:17 PM PDT
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Note, selecting 'dosemu2' instead of './dosemu2_2.0~pre9-1_amd64.deb'
The following additional packages will be installed:
  fluid-soundfont-gm
Suggested packages:
  valgrind install-otherdos fluid-soundfont-gs fluidsynth timidity
Recommended packages:
  install-freedos
The following NEW packages will be installed:
  dosemu2 fluid-soundfont-gm
0 upgraded, 2 newly installed, 0 to remove and 0 not upgraded.
Need to get 120 MB/122 MB of archives.
After this operation, 154 MB of additional disk space will be used.
```
After, `dpkg` shows `dosemu2` is fully installed.  I can type `dosemu` in a terminal and have a DOS window open up.

**Reply by stsp** [reply]
OK I removed a few things.
ubuntu builds will break again
but I don't think explicitly adding
llvm-13-linker-tools was the right
thing to do.

Not sure about bash because
AFAIK ubuntu installs dash by
default. So I left bash in for now.

Maybe @skitt knows if its safe
to remove bash from deps when
we explicitly need bash as /bin/sh?

**Reply by RossBoylan** [reply]
`/bin/sh` is a symlink to `dash` on my system; my understanding is that Debian has been doing it that way for awhile.  I wouldn't expect a dependency on `bash` by itself to change that.  I do have `bash` installed.

Regardless of `sh` being `dash` there didn't seem to be any problems as a result.

**Reply by stsp** [reply]
Sorry, not "as /bin/sh".
We explicitly write
`!#/bin/bash` when needed.
So I don't know why lintian
complains on that.

**Reply by skitt** [reply]
`bash` is an essential package, so it is always installed; that’s why lintian is complaining — there’s no need to depend on it, unless a specific version is required. https://github.com/dosemu2/dosemu2/pull/1693

**Reply by stsp** [reply]
Is it the case even on distros
that have dash as /bin/sh?
Why bash is essential for those?

**Reply by skitt** [reply]
Yes, this is the case on distros where the default `/bin/sh` is `dash` — both `dash` and `bash` are always installed. `bash` is essential in part for historical reasons — because it’s always been essential, packages which need it don’t depend on it; making `bash` non-essential would require first updating all packages using it to depend on it. In all likelihood, other essential packages would need `bash` anyway, making it transitively essential, so the exercise wouldn’t allow `bash` to be removed anyway. Any system with interactive users also needs `bash` since that’s the default login shell.

**Reply by stsp** [reply]
> Any system with interactive users also needs bash since that’s the default login shell.

Wow!
You mean, systems use dash
as /bin/sh, but still put `/bin/bash`
explicitly into /etc/passwd?
That sounds crazy.
The first thing I always do after
install, is to remove dash. :)
Whoever made it /bin/sh by
default, should be punished.

**Reply by stsp** [reply]
Also would you like to look
into https://bugs.launchpad.net/ubuntu/+source/llvm-toolchain-13/+bug/1979125
and possibly replicating it
into debian bugzilla?
I think ubuntu bugzilla is quite
hopeless unless debian has a fix.

**Reply by RossBoylan** [reply]
Lintian's complaint sounds as if it means that it is unnecessary to depend on essential packages, and `bash` and `coreutils` are essential packages. So lintian thinks the only reason for such a dependence would be to specify a version  If that's true, it's probably OK just to drop those dependencies.

**Reply by RossBoylan** [reply]
I think the main motive for using `dash` was to speed up boot times.

**Reply by skitt** [reply]
> > Any system with interactive users also needs bash since that’s the default login shell.
> 
> Wow! You mean, systems use dash as /bin/sh, but still put `/bin/bash` explicitly into /etc/passwd? That sounds crazy. The first thing I always do after install, is to remove dash. :) Whoever made it /bin/sh by default, should be punished.

Well, `dash` isn’t usable as an interactive shell. The idea behind making it the default `/bin/sh` is, as @RossBoylan says, for performance reasons — `dash` loads quite a bit faster than `bash`, so using it for `/bin/sh` shell scripts saves quite a bit of time. This was particularly significant back when initscripts were still in widespread use. (Incidentally, `bash` as `/bin/sh` is also not a great interactive shell, so systems where `bash` is `/bin/sh` also specify `/bin/bash` as their default shell, not `/bin/sh`.)

I’ll take a look at 1979125.

**Reply by RossBoylan** [reply]
> We explicitly write 
!#/bin/bash when needed.
So I don't know why lintian complains on that.

I doubt the messages are triggered by the `#!` lines. `dosemu2`'s `debian/control` file declares `build-depends:` and `depends:` on both `bash` and `coreutils`.  At least the version I built yesterday did.  The message seemed to be about the binary package, and so probably triggered just by the `Depends:` in the `Package: dosemu2` section at the bottom.

**Reply by stsp** [reply]
This is fixed.
